### PR TITLE
DCL software types in Kotlin DSL and DCL interpretation for `*.kts`

### PR DESCRIPTION
**Pull request by h0tk3y**
### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by h0tk3y** [comment]
@bot-gradle test this

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

File: platforms/core-configuration/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/KotlinDslDclSchemaCollector.kt [hunk 1]
```diff
@@ -16,36 +16,136 @@
...
+            TypeOf.typeOf<Any>(object : ParameterizedType {
+                override fun getActualTypeArguments(): Array<Type> = typeArgs.toTypedArray<Type>()
+                override fun getRawType(): Type = loadedClass
+
+                /** [Class.getNestHost] is @since 11, cannot use it; but we are fine with no owner type here. */
+                override fun getOwnerType() = null
+            })
+        } else{
+            TypeOf.typeOf(loadedClass)
         }
```
Comment (eskatos) [inline platforms/core-configuration/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/KotlinDslDclSchemaCollector.kt:144]: 💭  We could consider extracting this as a utility factory function in `TypeOf` for such cases

File: platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptPluginFactory.kt [hunk 2]
```diff
@@ -58,6 +70,20 @@ class KotlinScriptPluginFactory @Inject internal constructor(
...
+     * Given the multi-stage nature of DCL settings files, we must handle the cases
+     * when earlier stages like `pluginManagement` succeed but stages following them fail.
+     * One example strategy is to make Kotlin DSL avoid re-running the successful DCL stages
+     * and move right to the failed ones.
+     *
+     * That needs more changes in Kotlin DSL, and until we make them, we disable DCL interpretation
+     * for any target other than [Project] (which is single-stage in DCL and is thus safe).
+     */
+    private fun shouldTryDclInterpreterWithScriptTarget(target: Any) =
+        target is Project
```
Comment (eskatos) [inline platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptPluginFactory.kt:86]: 👍

**Review by eskatos — APPROVED** [review]
🎉

File: platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptPluginFactory.kt [hunk 3]
```diff
@@ -58,6 +69,8 @@ class KotlinScriptPluginFactory @Inject internal constructor(
     fun kotlinScriptOptions(): EnumSet<EvalOption> =
         if (inLenientMode()) lenientModeScriptOptions
         else defaultEvalOptions
+
+    private val logger = Logging.getLogger(KotlinScriptPluginFactory::class.java)
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptPluginFactory.kt:78]: Ah, right, the initial intention was to write some logs about the DCL interpretation. I added it along with the test in [`77583cd` (#31042)](https://github.com/gradle/gradle/pull/31042/commits/77583cd1e753eb1ff2375e46f3bf963d39c053c5).

**Comment by h0tk3y** [comment]
@bot-gradle test RFN

File: platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/provider/SettingsUnderInitialization.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import org.gradle.internal.event.ListenerManager
+import org.gradle.internal.service.scopes.Scope
+import org.gradle.internal.service.scopes.ServiceScope
+
+/**
+ * Captures an instance of [SettingsInternal] that is broadcast to [BuildListener] as early as [BuildListener.beforeSettings].
+ * Used as a workaround to inability to use [GradleInternal.getSettings] while evaluating a settings script.
+ */
+@ServiceScope(Scope.Build::class)
+internal class SettingsUnderInitialization(private val listenerManager: ListenerManager) : BuildListener, Stoppable {
```
[the remark this answers is not in the document]
Comment (h0tk3y) [reply platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/provider/SettingsUnderInitialization.kt]: Done in [`9d8daaf` (#31042)](https://github.com/gradle/gradle/pull/31042/commits/9d8daafab2f2daffb5c6332981de9e2c2343a5cf)

File: platforms/core-configuration/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/KotlinDslDclSchemaCollector.kt (hunk 1 above)
Comment (h0tk3y) [reply platforms/core-configuration/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/KotlinDslDclSchemaCollector.kt]: Done in [`d8d817d` (#31042)](https://github.com/gradle/gradle/pull/31042/commits/d8d817d2d726d844d4b4e667455914bcac2ae57e)

**Pull request merged**